pkg/container-utils/testutils/containerd.go

```diff
@@ -71,115 +109,170 @@ func RunContainerdContainer(ctx context.Context, t *testing.T, command string, o
-		if err != nil {
-			t.Fatalf("Failed to wait on task %q: %s", opts.name, err)
-		}
-		s := <-exitStatus
+	c.exitStatus, err = task.Wait(c.nsCtx)
+	if err != nil {
+		t.Fatalf("Failed to wait on task %q: %s", c.name, err)
+	}
+	c.pid = int(task.Pid())
+
+	if c.options.wait {
+		s := <-c.exitStatus
 		if s.ExitCode() != 0 {
-			t.Logf("Exitcode for task %q: %d", opts.name, s.ExitCode())
+			t.Logf("Exitcode for task %q: %d", c.name, s.ExitCode())
 		}
 	}
 
-	if opts.logs {
-		t.Logf("Container %q output:\n%s", opts.name, output.String())
+	if c.options.logs {
+		t.Logf("Container %q output:\n%s", c.name, output.String())
 	}
 
-	if opts.removal {
-		task.Kill(nsCtx, syscall.SIGKILL)
-		_, err = task.Delete(nsCtx)
+	if c.options.removal {
+		err := c.deleteAndClose(task, container)
 		if err != nil {
-			t.Fatalf("Failed to delete task %q: %s", opts.name, err)
-		}
-		err = container.Delete(nsCtx, containerd.WithSnapshotCleanup)
-		if err != nil {
-			t.Fatalf("Failed to delete container %q: %s", opts.name, err)
+			t.Fatalf("Failed to delete container %q: %s", c.name, err)
 		}
 	}
 }
 
-func RunContainerdFailedContainer(ctx context.Context, t *testing.T) {
-	RunContainerdContainer(ctx, t,
-		"/none",
-		WithName("test-ig-failed-container"),
-		WithoutLogs(),
-		WithoutWait(),
-	)
+func (c *ContainerdContainer) Start(t *testing.T) {
+	if c.started {
+		t.Logf("Warn(%s): trying to start already running container\n", c.name)
+		return
+	}
+	c.start(t)
+	c.started = true
 }
 
-func RemoveContainerdContainer(ctx context.Context, t *testing.T, name string) {
-	client, err := containerd.New("/run/containerd/containerd.sock",
-		containerd.WithTimeout(3*time.Second),
-	)
+func (c *ContainerdContainer) start(t *testing.T) {
+	for _, o := range []Option{WithoutWait(), WithoutRemoval()} {
+		o(c.options)
+	}
+	c.Run(t)
+}
+
+func (c *ContainerdContainer) Stop(t *testing.T) {
+	if !c.started && !c.options.forceDelete {
+		t.Logf("Warn(%s): trying to stop already stopped container\n", c.name)
+		return
+	}
+	if c.client == nil {
+		if c.options.forceDelete {
+			t.Logf("Warn(%s): trying to stop container with nil client. Forcing deletion\n", c.name)
+			if err := c.initClientAndCtx(); err != nil {
+				t.Fatalf("Failed to initialize client: %s", err)
+			}
+		} else {
+			t.Fatalf("Client is not initialized")
+		}
+	}
+
+	c.stop(t)
+	c.started = false
+}
+
+// deleteAndClose kill the task, delete the container and close the client
+func (c *ContainerdContainer) deleteAndClose(task containerd.Task, container containerd.Container) error {
+	var err error
+
+	// Kill the task
+	task.Kill(c.nsCtx, syscall.SIGKILL)
+
+	// We need to wait until the task is killed before trying to delete it
+	if c.options.forceDelete {
+		// If we are forcing the deletion, we don't want to wait forever as the
+		// task might be already stopped
+		select {
+		case <-c.exitStatus:
+		case <-time.After(1 * time.Second):
+		}
+	} else {
+		<-c.exitStatus
```

Comment:
Wouldn't the same issue you described 5 lines prior happen here too?

Reply:
You're right. It never happened to me, but it could happen